### ISSUE TITLE
Generate multiple `conanbuild.X` scripts from `VirtualBuildEnv` generator

### ISSUE THREAD

**Issue by dribeas** [conan-io/conan#13866]
When we run `conan install` and use the `VirtualBuildEnv` generator for local development (i.e. the goal is to build locally, not to create a new package) on windows it creates `conanbuild.bat` that can be used to setup the environment to run `cmake` (or any other tool).

The `tools.env.virtualenv:powershell` configuration option can be used to generate `conanbuild.ps1`, but there does not seem to be a way of generating `conanbuild.sh` on windows.

In my case, I'm used to developing locally with _ancient_ tools, a terminal, `vim`, `cmake` directly on the command line... which I do mostly on macos, sometimes on linux.  We have a multi-platform product, and sometimes I have to build/debug on a remote windows machine.  It would be convenient to be able to use `git-bash` on the remote windows machine rather than having to use `cmd`/`powershell`.

**Comment by memsharded** [comment]
Hi @dribeas 

Thanks for your comments.

When Conan is running inside a Windows subsystem (like msys2 bash) it is possible to tell it so it creates the right files:
```
$ conan install --tool-requires=cmake/3.23.5 -g VirtualBuildEnv -c tools.microsoft.bash:subsystem=msys2 -c tools.microsoft.bash:active=True
$ source conanbuild.sh
$ cmake --version
cmake version 3.23.5

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```
In general, it would be expected that those confs you can define them in your profile, to make it more convenient.
I have tested the above in git-bash, and it works. So the missing piece would be to classify ``git-bash``, what kind of subsystem is it?

**Comment by dribeas** [comment]
:)  I was afraid of setting `tools.microsoft.bash:active=True` as I did not understand what impact it would have.  This is sufficient.  As of the type of subsystem, I believe under the hood it's `msys2`. I will try with that and see if it works.

Thank you!

**Comment by memsharded** [comment]
Yeah, the meaning of ``tools.microsoft.bash:active=True`` is exactly: "I am already running on bash, you don't need to prepend ``bash -c ...`` to your commands" while the ``subsystem`` is to let it know that it should generate the ``.sh`` files.

**Comment by schwaerz** [comment]
For me this does not work. When I set those two options, the build method will fail with the following error:
```
'.' is not recognized as an internal or external command
```
Looks like it is still trying to use cmd to run the scripts. If I omit `tools.microsoft.bash:active=True` the build's working but still does not generate the desired `conanbuild.sh` file.

Will it be possible for conan to detect the environment and 'do the right thing' depending on the shell it is being executed from?

**Comment by memsharded** [comment]
> Will it be possible for conan to detect the environment and 'do the right thing' depending on the shell it is being executed from?

We moved practically all detection we had inline in Conan codebase in 1.X to explicit confs and user/profile inputs. They were problematic, reliably detecting the environment is not that easy and robust, and there are some use cases that such detection is not what the user wants. So explicit specification of the user is needed.

A different thing is that there still might be some cases and bugs to be improved, the truth is that the Windows subsystems are quite challenging. We need to check what is happening. For the above message.

> '.' is not recognized as an internal or external command

It seems that the shell is not able to ``source`` a shell script? Is this the case? Does it happens if you try to run ``source some.sh`` in the terminal.

**Comment by JoshMusick** [comment]
@memsharded 

I know this issues was last commented on in Sept, but I seem to be having a similar issue.  I can open a separate ticket if needed, but I think this may fit into this topic.

We created  a Conan 2.x recipe for gnuplot, which essentially downloads the zip/tarball and makes the binary available, because we have C++ programs that try to run commands such as 
```
    system("gnuplot --version");
```
or other methods that pass in the input file to be plotted (the --version is the simplest example).  We are using git-bash on windows and as such, the .bat files don't properly setup the PATH for the runtime program.  When I pass the two options that you mentioned (`-c tools.microsoft.bash:subsystem=msys2 -c tools.microsoft.bash:active=True`) it does in fact generate the conanrun.sh and conanbuild.sh files.   I can also run `source conanbuild.sh` and it will append the bin folder to the path as expected (sorta).  However, when we run the test_package which is trying to consume the gnuplot package, (using the options indicated -c tools.microsoft.bash:subsystem=msys2....), it reports the error 

```
======== Testing the package: Building ========
gnuplot/5.4.8 (test package): Calling build()
gnuplot/5.4.8 (test package): Running CMake.configure()
gnuplot/5.4.8 (test package): RUN: cmake -G "Visual Studio 16 2019" -DCMAKE_TOOLCHAIN_FILE="D:/Projects/rhd-conan-package-recipes/recipes-v2/gnuplot/all/test_package/build/msvc-192-x86_64-17-debug/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="D:/Projects/rhd-conan-package-recipes/recipes-v2/gnuplot/all/test_package" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" "D:\Projects\rhd-conan-package-recipes\recipes-v2\gnuplot\all\test_package"
'.' is not recognized as an internal or external command,
operable program or batch file.
```
So somewhere in the configure / build step of the test_package it is not able to run the ". <filename.sh>" which is in the conanrun.sh and conanbuild.sh.

When I run the commands in a separate sandbox, with a simple conanfile.txt, I can successfully run `conan install ...` and then `source conanbuild.sh` and it will run the script without error.  But it is not currently appending the path in the correct format.  When I echo my path (from the git-bash window) prior to sourcing the file it looks like this ...

```
$ echo $PATH
/c/Users/jmusick/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/mingw64/bin:/usr/bin:...
```
But after sourcing the conanbuild.sh file...
```
$ source build/generators/conanbuild.sh
$ echo $PATH
C:\Users\jmusick\.conan2\p\b\gnuplf123b2fa56c07\p\bin /c/Users/jmusick/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/mingw64/bin:/usr/bin:...
```
It is mixing the "linux style" paths with windows style paths in the export command.

When I deactivate the environment, and change the command that is in the conanbuildenv-debug....sh file from 
```
export PATH="C:\Users\jmusick\.conan2\p\b\gnuplf123b2fa56c07\p\bin $PATH"
```
to the following:
```
export PATH="/c/Users/jmusick/.conan2/p/b/gnuplf123b2fa56c07/p/bin:$PATH"
```

Then running the command `gnuplot --version` works correctly.

Finally, I was trying to use the conan run environment, but no matter what I have tried, it does not add the path data to the `conanrunenv-debug.....sh` file, it only adds to the `conanbuildenv-debug....sh`.   

I have tried adding the following to the gnuplot recipe package_info method, but it only adds the same path twice.
```
    pt = os.path.join(self.package_folder, self.cpp_info.bindirs[0])
    self.runenv_info.append("PATH", pt)
```

My specific issues seem to be: 
1. The path being appended when running a windows bash is not properly converting it to the linux style path
2. When running the test package it is not properly using a bash shell, but perhaps it is still running windows cmd.exe (which is why it complains about the "`.` is not recognized"
3. Passing information to the conanrun environment does not seem to take effect, it only ever gets sent to the conanbuldenv.

Thanks again!
v/r
Josh
